### PR TITLE
FastCV-based HAL for OpenCV acceleration 2ndpost-3

### PR DESCRIPTION
**Pull request by quic-apreetam**
### Detailed description:

- Add cv_hal_canny for Canny API

Requires binary from [opencv/opencv_3rdparty#90](https://github.com/opencv/opencv_3rdparty/pull/90) 
Depends on: [opencv/opencv#26617](https://github.com/opencv/opencv/pull/26617)
Depends on: [opencv/opencv#26619](https://github.com/opencv/opencv/pull/26619) 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: 3rdparty/fastcv/src/fastcv_hal_core.cpp [hunk 1]
```diff
@@ -324,11 +324,251 @@ int fastcv_hal_rotate(
...
+    size_t          dst_step,
+    int             width,
+    int             height,
+    double          scale)
+{
+    int8_t sF;
+
+    if(scale >= 1.0)
+    {
+        if(scale == 1.0)        { sF =  0; }
```
Comment (savuor) [inline 3rdparty/fastcv/src/fastcv_hal_core.cpp:403]: Here and below:
Float point comparisons should be done with epsilon, otherwise they are unpredictable

Comment (quic-apreetam) [reply 3rdparty/fastcv/src/fastcv_hal_core.cpp:403]: thanks @savuor, will be updated

**Review by asmorkalov — APPROVED** [review]

**Comment by asmorkalov** [comment]
@quic-apreetam The first part has been merged. Please rebase your patch and fix conflicts. The patch provides Canny only. I expect, the PR should be independent from post-2.

**Comment by quic-apreetam** [comment]
> @quic-apreetam The first part has been merged. Please rebase your patch and fix conflicts. The patch provides Canny only. I expect, the PR should be independent from post-2.

hi @asmorkalov, @adsha-quic will have to rebase https://github.com/opencv/opencv/pull/26619 again if I resolve the conflict now. Once https://github.com/opencv/opencv/pull/26619 is merged, I will resolve the conflict.

**Comment by asmorkalov** [comment]
The second part has been merged. Please rebase the PR on top of 4.x.

**Comment by quic-apreetam** [comment]
> The second part has been merged. Please rebase the PR on top of 4.x.

hi @asmorkalov rebased on top 4.x and enabled fastcv by default.

File: 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp [hunk 2]
```diff
@@ -992,4 +992,60 @@ int fastcv_hal_cvtBGRtoYUVApprox(
...
+        CV_HAL_RETURN_NOT_IMPLEMENTED("Multi-channel input is not supported");
+
+    if (lowThreshold > highThreshold)
+        CV_HAL_RETURN_NOT_IMPLEMENTED("lowThreshold is greater then highThreshold");
+
+    double	lowThresh = lowThreshold;
+    double	highThresh = highThreshold;
+
+    if (lowThresh != static_cast<int>(lowThresh) || highThresh != static_cast<int>(highThresh))
+        CV_HAL_RETURN_NOT_IMPLEMENTED("threshold with decimal values not supported");
```
Comment (asmorkalov) [inline 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp:1025]: 1. Looks like lowThresh and highThresh are redundant.
2. What if, lowThresh=12.999999999? I propose to use check with epsilon.

Comment (quic-apreetam) [reply 3rdparty/fastcv/src/fastcv_hal_imgproc.cpp]: > 1. Looks like lowThresh and highThresh are redundant.
> 2. What if, lowThresh=12.999999999? I propose to use check with epsilon.

hi @asmorkalov, updated

**Comment by asmorkalov** [comment]
The PR passed Android CI. Please revert root cmake changes and I'll merge it.

**Pull request merged**